lib/queue.js

```diff
@@ -1,256 +1,497 @@
-var redis = require('redis');
-var events = require('events');
-var util = require('util');
+'use strict';
 
-var Job = require('./job');
-var defaults = require('./defaults');
-var lua = require('./lua');
-var helpers = require('./helpers');
-var barrier = helpers.barrier;
+const redis = require('./redis');
+const Emitter = require('events').EventEmitter;
+const util = require('util');
+
+const Job = require('./job');
+const defaults = require('./defaults');
+const lua = require('./lua');
+const helpers = require('./helpers');
+const backoff = require('./backoff');
+const EagerTimer = require('./eager-timer');
 
 function Queue(name, settings) {
   if (!(this instanceof Queue)) {
     return new Queue(name, settings);
   }
 
+  Emitter.call(this);
+
   this.name = name;
   this.paused = false;
-  this.jobs = {};
+  this.jobs = new Map();
+  this.activeJobs = new Set();
+  this.checkTimer = null;
+
+  this._closed = null;
+  this._isClosed = false;
 
   settings = settings || {};
   this.settings = {
     redis: settings.redis || {},
-    stallInterval: typeof settings.stallInterval === 'number' ?
-      settings.stallInterval :
-      defaults.stallInterval,
     keyPrefix: (settings.prefix || defaults.prefix) + ':' + this.name + ':'
   };
 
-  var boolProps = ['isWorker', 'getEvents', 'sendEvents', 'removeOnSuccess', 'catchExceptions'];
-  boolProps.forEach(function (prop) {
-    this.settings[prop] = typeof settings[prop] === 'boolean' ? settings[prop] : defaults[prop];
-  }.bind(this));
+  for (let prop in defaults) {
+    const def = defaults[prop], setting = settings[prop], type = typeof def;
+    if (type === 'boolean') {
+      this.settings[prop] = typeof setting === 'boolean' ? setting : def;
+    } else if (type === 'number') {
+      this.settings[prop] = Number.isSafeInteger(setting) ? setting : def;
+    }
+  }
 
   /* istanbul ignore if */
   if (this.settings.redis.socket) {
-    this.settings.redis.params = [this.settings.redis.socket, this.settings.redis.options];
-  } else {
-    this.settings.redis.port = this.settings.redis.port || 6379;
-    this.settings.redis.host = this.settings.redis.host || '127.0.0.1';
-    this.settings.redis.params = [
-      this.settings.redis.port, this.settings.redis.host, this.settings.redis.options
-    ];
+    this.settings.redis = Object.assign({}, this.settings.redis, {
+      path: this.settings.redis.socket
+    });
   }
-  this.settings.redis.db = this.settings.redis.db || 0;
 
-  // Wait for Lua loading and client connection; bclient and eclient/subscribe if needed
-  var reportReady = barrier(
-    2 + this.settings.isWorker + this.settings.getEvents * 2,
-    this.emit.bind(this, 'ready')
-  );
+  this._delayedTimer = new EagerTimer(this.settings.nearTermWindow);
```

Comment:
I guess this is fine behavior-wise since `_processDelayed` early-exits if `options.processDelayed` is false, but it seems weird to create this timer if we know it'll never do anything. The only place that touches the timer when `options.processDelayed` is false is the `close()` cleanup stopping it, so I think it would be reasonable to only create the timer if `options.processDelayed` is true.